src/App.tsx

```diff
@@ -1,24 +1,223 @@
+  const [processingTodoIds, setProcessingTodoIds] = useState<number[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const hasTodos = todos.length > 0;
+
+  const isAllCompleted = todos.every(todo => todo.completed);
+
+  const fetchData = async () => {
+    try {
+      setErrorMessage(null);
+      const todosFetch = await todoService.getTodos();
+
+      setTodos(todosFetch);
+    } catch (err) {
+      setErrorMessage('Unable to load todos');
+    }
+  };
+
+  useEffect(() => {
+    fetchData();
+  }, []);
+
+  const hadleAddTodo = (title: string) => {
+    setInputDisabled(true);
+
+    setTempTodo({
+      id: 0,
+      userId: 0,
+      title: title.trim(),
+      completed: false,
+    });
+
+    return todoService
+      .addTodo(title)
+      .then((newTodo) => {
+        setTodos((prevTodos) => [...prevTodos, newTodo]);
+      })
+      .catch((error) => {
+        setErrorMessage('Unable to add a todo');
+        throw error;
+      })
+      .finally(() => {
+        setInputDisabled(false);
+        setTempTodo(null);
+      });
+  };
+
+  const filteredTodos = useMemo(() => todos.filter(({ completed }) => {
+    switch (condition) {
+      case ForComletedTodo.Active:
+        return !completed;
+      case ForComletedTodo.Completed:
+        return completed;
+      default:
+        return 1;
+    }
+  }), [condition, todos]);
+
+  const handleDeleteTodo = (todoId: number) => {
+    setInputDisabled(true);
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todoId]);
+
+    return todoService
+      .deleteTodo(todoId)
+      .then((() => {
+        setTodos((prevTodos) => prevTodos.filter(todo => todo.id !== todoId));
+      }))
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+        throw new Error();
+      })
+      .finally(() => {
+        setInputDisabled(false);
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todoId),
+        );
+      });
+  };
+
+  const handleRenameTodo = async (todo: Todo, newTodoTitle: string) => {
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todo.id]);
+
+    return todoService.updateTodo({
+      id: todo.id,
+      title: newTodoTitle.trim(),
+      userId: todo.userId,
+      completed: todo.completed,
+    })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map(currentTodo => (
+          currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo
+        )));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to update a todo');
+        throw new Error();
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todo.id),
+        );
+      });
+  };
+
+  const handleClearCompletedTodos = () => {
+    todos
+      .filter(todo => todo.completed)
+      .forEach(todo => {
+        handleDeleteTodo(todo.id);
+      });
```

Comment:
It is not necessary to go through the array twice, it is enough to transfer the filtering as a condition to the middle of the `forEach`.